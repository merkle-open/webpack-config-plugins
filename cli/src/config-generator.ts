export interface ConfigLoaderTypeOptions {
	useJs: boolean;
	useTs: boolean;
	useScss: boolean;
	useCss: boolean;
	useFonts: boolean;
	useImages: boolean;
	useHtml: boolean;
	useClean: boolean;
	useDevServer: boolean;
	useCli: boolean;
}

export type GeneratedConfigs = {
	npmInstall: string;
	webpackConfig: string;
};

export type ConfigOptionKeys = keyof ConfigLoaderTypeOptions;

type PluginName =
	| 'js-config-webpack-plugin'
	| 'ts-config-webpack-plugin'
	| 'scss-config-webpack-plugin'
	| 'font-config-webpack-plugin'
	| 'image-config-webpack-plugin'
	| 'common-config-webpack-plugin'
	| 'asset-config-webpack-plugin'
	| 'html-webpack-plugin'
	| 'clean-webpack-plugin'
	| 'webpack-dev-server'
	| 'webpack-cli';

const moduleNames: { [option in ConfigOptionKeys]: PluginName } = {
	useJs: 'js-config-webpack-plugin',
	useTs: 'ts-config-webpack-plugin',
	useScss: 'scss-config-webpack-plugin',
	useCss: 'scss-config-webpack-plugin',
	useFonts: 'font-config-webpack-plugin',
	useImages: 'image-config-webpack-plugin',
	useHtml: 'html-webpack-plugin',
	useClean: 'clean-webpack-plugin',
	useDevServer: 'webpack-dev-server',
	useCli: 'webpack-cli',
};

export const configOptionKeys = Object.keys(moduleNames) as Array<ConfigOptionKeys>;

const assetConfigWebpackPluginChildren: Array<PluginName> = [
	'font-config-webpack-plugin',
	'image-config-webpack-plugin',
];

const commonConfigWebpackPluginChildren: Array<PluginName> = [
	'js-config-webpack-plugin',
	'ts-config-webpack-plugin',
	'scss-config-webpack-plugin',
	'asset-config-webpack-plugin',
];

const dependencies: { [option in PluginName]: Array<string> } = {
	'js-config-webpack-plugin': ['js-config-webpack-plugin', '@babel/core', '@babel/preset-env', '@babel/preset-react'],
	'ts-config-webpack-plugin': ['ts-config-webpack-plugin', 'typescript'],
	'scss-config-webpack-plugin': ['scss-config-webpack-plugin'],
	'font-config-webpack-plugin': ['font-config-webpack-plugin'],
	'image-config-webpack-plugin': ['image-config-webpack-plugin'],
	'common-config-webpack-plugin': ['common-config-webpack-plugin'],
	'asset-config-webpack-plugin': ['asset-config-webpack-plugin'],
	'html-webpack-plugin': ['html-webpack-plugin'],
	'clean-webpack-plugin': ['clean-webpack-plugin'],
	'webpack-dev-server': ['webpack-dev-server', 'webpack'],
	'webpack-cli': ['webpack', 'webpack-cli'],
};

const webpackPluginImports: { [option in PluginName]: string | false } = {
	'js-config-webpack-plugin': `const JsConfigWebpackPlugin = require('js-config-webpack-plugin')`,
	'ts-config-webpack-plugin': `const TsConfigWebpackPlugin = require('ts-config-webpack-plugin')`,
	'scss-config-webpack-plugin': `const ScssConfigWebpackPlugin = require('scss-config-webpack-plugin')`,
	'font-config-webpack-plugin': `const FontsConfigWebpackPlugin = require('font-config-webpack-plugin')`,
	'image-config-webpack-plugin': `const ImagesConfigWebpackPlugin = require('image-config-webpack-plugin')`,
	'common-config-webpack-plugin': `const CommonConfigWebpackPlugin = require('common-config-webpack-plugin')`,
	'asset-config-webpack-plugin': `const AssetConfigWebpackPlugin = require('asset-config-webpack-plugin')`,
	'html-webpack-plugin': `const HtmlWebpackPlugin = require('html-webpack-plugin')`,
	'clean-webpack-plugin': `const { CleanWebpackPlugin } = require('clean-webpack-plugin').CleanWebpackPlugin`,
	'webpack-dev-server': false,
	'webpack-cli': false,
};

const webpackPluginConfigurations: { [option in PluginName]: string | false } = {
	'js-config-webpack-plugin': 'new JsConfigWebpackPlugin()',
	'ts-config-webpack-plugin': 'new TsConfigWebpackPlugin()',
	'scss-config-webpack-plugin': 'new ScssConfigWebpackPlugin()',
	'font-config-webpack-plugin': 'new FontsConfigWebpackPlugin()',
	'image-config-webpack-plugin': 'new ImagesConfigWebpackPlugin()',
	'common-config-webpack-plugin': 'new CommonConfigWebpackPlugin()',
	'asset-config-webpack-plugin': 'new AssetConfigWebpackPlugin()',
	'html-webpack-plugin': 'new HtmlWebpackPlugin()',
	'clean-webpack-plugin': `new CleanWebpackPlugin()`,
	'webpack-dev-server': false,
	'webpack-cli': false,
};

const webpackPluginExplanation: { [option in PluginName]: Array<string> | false } = {
	'js-config-webpack-plugin': [
		'Multi threading babel loader configuration with caching for .js and .jsx files',
		'see https://github.com/namics/webpack-config-plugins/tree/master/packages/js-config-webpack-plugin/config',
	],
	'ts-config-webpack-plugin': [
		'Multi threading typescript loader configuration with caching for .ts and .tsx files',
		'see https://github.com/namics/webpack-config-plugins/tree/master/packages/ts-config-webpack-plugin/config',
	],
	'scss-config-webpack-plugin': [
		'SCSS Configuration for .css .module.css and .scss .module.scss files',
		'see https://github.com/namics/webpack-config-plugins/tree/master/packages/scss-config-webpack-plugin/config',
	],
	'font-config-webpack-plugin': [
		'File loader configuration for .woff and .woff2 files',
		'see https://github.com/namics/webpack-config-plugins/tree/master/packages/font-config-webpack-plugin/config',
	],
	'image-config-webpack-plugin': [
		'File loader configuration for .gif .jpg .jpeg .png and .svg files',
		'see https://github.com/namics/webpack-config-plugins/tree/master/packages/image-config-webpack-plugin/config',
	],
	'common-config-webpack-plugin': [
		'Multi threading babel loader configuration with caching for .js and .jsx files',
		'Multi threading typescript loader configuration with caching for .ts and .tsx files',
		'SCSS Configuration for .css .module.css and .scss .module.scss files',
		'File loader configuration for .woff and .woff2 files',
		'File loader configuration for .gif .jpg .jpeg .png and .svg files',
		'https://github.com/namics/webpack-config-plugins/',
	],
	'asset-config-webpack-plugin': [
		'File loader configuration for .woff and .woff2 files',
		'File loader configuration for .gif .jpg .jpeg .png and .svg files',
		'https://github.com/namics/webpack-config-plugins/tree/master/packages/asset-config-webpack-plugin',
	],
	'html-webpack-plugin': ['Generate a base html file and injects all generated css and js files'],
	'clean-webpack-plugin': ['Cleans the dist folder before the build starts'],
	'webpack-dev-server': false,
	'webpack-cli': false,
};

export const configOptionLabels: { [fieldName in keyof ConfigLoaderTypeOptions]: string } = {
	useJs: 'Load ES6 (.js .jsx)',
	useTs: 'Load Typescript (.ts tsx)',
	useScss: 'Load Scss (.scss)',
	useCss: 'Load Css (.css)',
	useFonts: 'Load Fonts (.woff .woff2)',
	useImages: 'Load Images',
	useHtml: 'Use Html Plugin',
	useClean: 'Use Clean Plugin',
	useDevServer: 'webpack-dev-server',
	useCli: 'webpack-cli',
};

/**
 * Returns the module names for the given config
 *
 * e.g. getModulesForConfiguration({useJs: true}) // -> 'js-config-webpack-plugin'
 *
 * common-config-webpack-plugin
 * ├── js-config-webpack-plugin
 * ├── ts-config-webpack-plugin
 * ├── scss-config-webpack-plugin
 * └── asset-config-webpack-plugin
 *     ├── font-config-webpack-plugin
 *     └── image-config-webpack-plugin
 */
function getModuleNamesForConfiguration(configOptions: ConfigLoaderTypeOptions): Array<PluginName> {
	return (Object.keys(configOptions) as Array<keyof typeof configOptions>)
		.filter((key) => configOptions[key])
		.map((key) => moduleNames[key]);
}

/**
 * Reduce the plugins by using wrappers
 */
function combinePluginsToWrappers(moduleNamesForConfiguration: Array<PluginName>): Array<PluginName> {
	// Use Asset Plugin
	const useAssetConfigWebpackPlugin = !assetConfigWebpackPluginChildren.some(
		(moduleName) => moduleNamesForConfiguration.indexOf(moduleName) === -1
	);
	if (useAssetConfigWebpackPlugin) {
		moduleNamesForConfiguration = moduleNamesForConfiguration.filter(
			(moduleName) => assetConfigWebpackPluginChildren.indexOf(moduleName) === -1
		);
		moduleNamesForConfiguration.push('asset-config-webpack-plugin');
	}
	// Use common config plugin
	const useCommonConfigWebpackPlugin = !commonConfigWebpackPluginChildren.some(
		(moduleName) => moduleNamesForConfiguration.indexOf(moduleName) === -1
	);
	if (useCommonConfigWebpackPlugin) {
		moduleNamesForConfiguration = moduleNamesForConfiguration.filter(
			(moduleName) => commonConfigWebpackPluginChildren.indexOf(moduleName) === -1
		);
		moduleNamesForConfiguration.push('common-config-webpack-plugin');
	}
	return moduleNamesForConfiguration;
}

function unqiue<U extends string, T extends Array<any> = Array<U>>(array: T): T {
	return Array.from(new Set(array).keys()) as T;
}

export function generateConfigurations(configOptions: ConfigLoaderTypeOptions): GeneratedConfigs {
	const moduleNames = unqiue(combinePluginsToWrappers(getModuleNamesForConfiguration(configOptions)));
	moduleNames.sort();
	const dependencsToInstall: Array<string> = [];
	moduleNames.forEach((moduleName) => {
		dependencsToInstall.push(...dependencies[moduleName]);
	});

	const uniqueDependencsToInstall = unqiue(dependencsToInstall);
	uniqueDependencsToInstall.sort();

	const pluginImports = moduleNames
		.map((moduleName) => webpackPluginImports[moduleName])
		.filter((value) => value) as Array<string>;

	const pluginConfigurations = moduleNames
		.filter((moduleName) => webpackPluginConfigurations[moduleName])
		.map((moduleName) => {
			const moduleExplanation = webpackPluginExplanation[moduleName];
			const explanation =
				moduleExplanation === false ? '' : '// ' + moduleExplanation.join('\n    // ') + '\n    ';
			return explanation + webpackPluginConfigurations[moduleName];
		}) as Array<string>;

	return {
		npmInstall: uniqueDependencsToInstall.length ? `${[''].concat(uniqueDependencsToInstall).join(' ')}` : '',
		webpackConfig: pluginImports.length
			? `${pluginImports.join(';\n')};

module.exports = {
  plugins: [
    ${pluginConfigurations.join(',\n    ')},
  ],
};
`
			: '',
	};
}
