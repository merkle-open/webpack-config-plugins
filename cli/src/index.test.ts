import * as inquirer from 'inquirer';
import * as path from 'path';
import * as fs from 'fs';
import * as rimraf from 'rimraf';
import { generateConfigCli } from '.';
import { ConfigOptionKeys } from './config-generator';

// Fix invalid mockFs typings
jest.mock('inquirer');

let folderIndex = 0;
function generateTmpFolder(name = String(folderIndex++)) {
	const tmpFolder = path.resolve(__dirname, '../tmp');
	const tmpSubFolder = path.join(tmpFolder, name);
	rimraf.sync(tmpSubFolder);
	try {
		fs.mkdirSync(tmpFolder);
	} catch (e) {}
	try {
		fs.mkdirSync(tmpSubFolder);
	} catch (e) {}
	return tmpSubFolder;
}

function mergeLoggerMockCalls(fn: ReturnType<typeof jest.fn>) {
	return fn.mock.calls.reduce((result, call) => result + call.join(' ') + '\n', '\n');
}

type InquirerQuestions = { options: Array<ConfigOptionKeys>; overwrite: 'yes' | 'no'; install: 'yes' | 'no' };
const mockInquirer = (mockAnswers: Partial<InquirerQuestions>) => {
	inquirer.prompt.mockImplementation(async (questions: Array<{ name: string }>) => {
		if (questions.length !== 1) {
			throw new Error('Can only mock single questions');
		}
		const answer = (mockAnswers as any)[questions[0].name];
		if (typeof answer === undefined) {
			throw new Error('Missing mock for ' + questions[0].name);
		}
		return {
			[questions[0].name]: answer,
		};
	});
};

// Combinations to test
// prettier-ignore
const combinations: Array<{ options: Array<ConfigOptionKeys> }> = [
	{ options: ['useClean','useCli','useCss','useDevServer','useFonts','useHtml','useImages','useJs','useScss','useTs'] },
	{ options: ['useClean','useCli','useCss','useDevServer','useFonts','useHtml','useImages','useScss','useTs'] },
	{ options: ['useClean','useCli','useCss','useDevServer','useFonts','useHtml','useImages','useScss','useJs'] },
	{ options: ['useCli', 'useTs'] },
	{ options: ['useCli', 'useJs'] },
	{ options: ['useDevServer', 'useTs'] },
	{ options: ['useDevServer', 'useJs'] },
	{ options: ['useDevServer', 'useTs', 'useImages', 'useScss'] },
	{ options: ['useDevServer', 'useJs', 'useImages', 'useScss'] },
];

describe('cli', () => {
	test.each(combinations as any)(
		`it should run without errors for options %j`,
		async ({ options }: { options: Array<ConfigOptionKeys> }) => {
			const logger = jest.fn();
			console.log = logger;
			mockInquirer({ options, overwrite: 'no', install: 'no' });
			await generateConfigCli(process.cwd());
			expect(logger.mock.calls.length).not.toBe(0);
		}
	);

	test.each(combinations as any)(
		`it output the correct webpack.config.js content for %j`,
		async ({ options }: { options: Array<ConfigOptionKeys> }) => {
			const logger = jest.fn();
			console.log = logger;
			mockInquirer({ options, overwrite: 'no', install: 'no' });
			await generateConfigCli(process.cwd());
			expect(mergeLoggerMockCalls(logger)).toMatchSnapshot();
		}
	);

	test.each(combinations as any)(
		`it logs a understandable instruction if webpack.config.js content is written %j`,
		async ({ options }: { options: Array<ConfigOptionKeys> }) => {
			const tmpSubFolder = generateTmpFolder(options.join('-'));
			const logger = jest.fn();
			console.log = logger;
			mockInquirer({ options, overwrite: 'yes', install: 'no' });
			await generateConfigCli(tmpSubFolder);
			expect(mergeLoggerMockCalls(logger)).toMatchSnapshot();
		}
	);
});
