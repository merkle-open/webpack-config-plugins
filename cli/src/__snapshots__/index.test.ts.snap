// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli it logs a understandable instruction if webpack.config.js content is written {"options":["useClean","useCli","useCss","useDevServer","useFonts","useHtml","useImages","useJs","useScss","useTs"]} 1`] = `
"
✍️  Creating \\"webpack.config.js\\"
⚠️  Skipping NPM installation
💡  To install the dependencies manually run:

npm install --save-dev  clean-webpack-plugin common-config-webpack-plugin html-webpack-plugin webpack webpack-cli webpack-dev-server


Next steps:
💡  Webpack will look for \\"src/index.ts\\" or \\"src/index.tsx\\" which you have to create manually.
💡  Build your bundle with    webpack --mode production
💡  Start the server with     webpack-dev-server --mode development
💡  For more information on avaialble modes please go to https://webpack.js.org/concepts/mode/
"
`;

exports[`cli it logs a understandable instruction if webpack.config.js content is written {"options":["useClean","useCli","useCss","useDevServer","useFonts","useHtml","useImages","useScss","useJs"]} 1`] = `
"
✍️  Creating \\"webpack.config.js\\"
⚠️  Skipping NPM installation
💡  To install the dependencies manually run:

npm install --save-dev  @babel/core @babel/preset-env @babel/preset-react asset-config-webpack-plugin clean-webpack-plugin html-webpack-plugin js-config-webpack-plugin scss-config-webpack-plugin webpack webpack-cli webpack-dev-server


Next steps:
💡  Webpack will look for \\"src/index.js\\" which you have to create manually.
💡  Build your bundle with    webpack --mode production
💡  Start the server with     webpack-dev-server --mode development
💡  For more information on avaialble modes please go to https://webpack.js.org/concepts/mode/
"
`;

exports[`cli it logs a understandable instruction if webpack.config.js content is written {"options":["useClean","useCli","useCss","useDevServer","useFonts","useHtml","useImages","useScss","useTs"]} 1`] = `
"
✍️  Creating \\"webpack.config.js\\"
⚠️  Skipping NPM installation
💡  To install the dependencies manually run:

npm install --save-dev  asset-config-webpack-plugin clean-webpack-plugin html-webpack-plugin scss-config-webpack-plugin ts-config-webpack-plugin typescript webpack webpack-cli webpack-dev-server


Next steps:
💡  Webpack will look for \\"src/index.ts\\" or \\"src/index.tsx\\" which you have to create manually.
💡  Build your bundle with    webpack --mode production
💡  Start the server with     webpack-dev-server --mode development
💡  For more information on avaialble modes please go to https://webpack.js.org/concepts/mode/
"
`;

exports[`cli it logs a understandable instruction if webpack.config.js content is written {"options":["useCli","useJs"]} 1`] = `
"
✍️  Creating \\"webpack.config.js\\"
⚠️  Skipping NPM installation
💡  To install the dependencies manually run:

npm install --save-dev  @babel/core @babel/preset-env @babel/preset-react js-config-webpack-plugin webpack webpack-cli


Next steps:
💡  Webpack will look for \\"src/index.js\\" which you have to create manually.
💡  Build your bundle with    webpack --mode production
💡  For more information on avaialble modes please go to https://webpack.js.org/concepts/mode/
"
`;

exports[`cli it logs a understandable instruction if webpack.config.js content is written {"options":["useCli","useTs"]} 1`] = `
"
✍️  Creating \\"webpack.config.js\\"
⚠️  Skipping NPM installation
💡  To install the dependencies manually run:

npm install --save-dev  ts-config-webpack-plugin typescript webpack webpack-cli


Next steps:
💡  Webpack will look for \\"src/index.ts\\" or \\"src/index.tsx\\" which you have to create manually.
💡  Build your bundle with    webpack --mode production
💡  For more information on avaialble modes please go to https://webpack.js.org/concepts/mode/
"
`;

exports[`cli it logs a understandable instruction if webpack.config.js content is written {"options":["useDevServer","useJs","useImages","useScss"]} 1`] = `
"
✍️  Creating \\"webpack.config.js\\"
⚠️  Skipping NPM installation
💡  To install the dependencies manually run:

npm install --save-dev  @babel/core @babel/preset-env @babel/preset-react image-config-webpack-plugin js-config-webpack-plugin scss-config-webpack-plugin webpack webpack-dev-server


Next steps:
💡  Webpack will look for \\"src/index.js\\" which you have to create manually.
💡  Start the server with     webpack-dev-server --mode development
💡  For more information on avaialble modes please go to https://webpack.js.org/concepts/mode/
"
`;

exports[`cli it logs a understandable instruction if webpack.config.js content is written {"options":["useDevServer","useJs"]} 1`] = `
"
✍️  Creating \\"webpack.config.js\\"
⚠️  Skipping NPM installation
💡  To install the dependencies manually run:

npm install --save-dev  @babel/core @babel/preset-env @babel/preset-react js-config-webpack-plugin webpack webpack-dev-server


Next steps:
💡  Webpack will look for \\"src/index.js\\" which you have to create manually.
💡  Start the server with     webpack-dev-server --mode development
💡  For more information on avaialble modes please go to https://webpack.js.org/concepts/mode/
"
`;

exports[`cli it logs a understandable instruction if webpack.config.js content is written {"options":["useDevServer","useTs","useImages","useScss"]} 1`] = `
"
✍️  Creating \\"webpack.config.js\\"
⚠️  Skipping NPM installation
💡  To install the dependencies manually run:

npm install --save-dev  image-config-webpack-plugin scss-config-webpack-plugin ts-config-webpack-plugin typescript webpack webpack-dev-server


Next steps:
💡  Webpack will look for \\"src/index.ts\\" or \\"src/index.tsx\\" which you have to create manually.
💡  Start the server with     webpack-dev-server --mode development
💡  For more information on avaialble modes please go to https://webpack.js.org/concepts/mode/
"
`;

exports[`cli it logs a understandable instruction if webpack.config.js content is written {"options":["useDevServer","useTs"]} 1`] = `
"
✍️  Creating \\"webpack.config.js\\"
⚠️  Skipping NPM installation
💡  To install the dependencies manually run:

npm install --save-dev  ts-config-webpack-plugin typescript webpack webpack-dev-server


Next steps:
💡  Webpack will look for \\"src/index.ts\\" or \\"src/index.tsx\\" which you have to create manually.
💡  Start the server with     webpack-dev-server --mode development
💡  For more information on avaialble modes please go to https://webpack.js.org/concepts/mode/
"
`;

exports[`cli it output the correct webpack.config.js content for {"options":["useClean","useCli","useCss","useDevServer","useFonts","useHtml","useImages","useJs","useScss","useTs"]} 1`] = `
"
⚠️ Skipping writing webpack.config.js
💡  Add the following code to your webpack.config.js:

const CleanWebpackPlugin = require('clean-webpack-plugin');
const CommonConfigWebpackPlugin = require('common-config-webpack-plugin');
const HtmlWebpackPlugin = require('html-webpack-plugin');

module.exports = {
  plugins: [
    // Cleans the dist folder before the build starts
    new CleanWebpackPlugin(['dist']),
    // Multi threading babel loader configuration with caching for .js and .jsx files
    // Multi threading typescript loader configuration with caching for .ts and .tsx files
    // SCSS Configuration for .css .module.css and .scss .module.scss files
    // File loader configuration for .woff and .woff2 files
    // File loader configuration for .gif .jpg .jpeg .png and .svg files
    // https://github.com/namics/webpack-config-plugins/
    new CommonConfigWebpackPlugin(),
    // Generate a base html file and injects all generated css and js files
    new HtmlWebpackPlugin(),
  ],
};


⚠️  Skipping NPM installation
💡  To install the dependencies manually run:

npm install --save-dev  clean-webpack-plugin common-config-webpack-plugin html-webpack-plugin webpack webpack-cli webpack-dev-server

"
`;

exports[`cli it output the correct webpack.config.js content for {"options":["useClean","useCli","useCss","useDevServer","useFonts","useHtml","useImages","useScss","useJs"]} 1`] = `
"
⚠️ Skipping writing webpack.config.js
💡  Add the following code to your webpack.config.js:

const AssetConfigWebpackPlugin = require('asset-config-webpack-plugin');
const CleanWebpackPlugin = require('clean-webpack-plugin');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const JsConfigWebpackPlugin = require('js-config-webpack-plugin');
const ScssConfigWebpackPlugin = require('scss-config-webpack-plugin');

module.exports = {
  plugins: [
    // File loader configuration for .woff and .woff2 files
    // File loader configuration for .gif .jpg .jpeg .png and .svg files
    // https://github.com/namics/webpack-config-plugins/tree/master/packages/asset-config-webpack-plugin
    new AssetConfigWebpackPlugin(),
    // Cleans the dist folder before the build starts
    new CleanWebpackPlugin(['dist']),
    // Generate a base html file and injects all generated css and js files
    new HtmlWebpackPlugin(),
    // Multi threading babel loader configuration with caching for .js and .jsx files
    // see https://github.com/namics/webpack-config-plugins/tree/master/packages/js-config-webpack-plugin/config
    new JsConfigWebpackPlugin(),
    // SCSS Configuration for .css .module.css and .scss .module.scss files
    // see https://github.com/namics/webpack-config-plugins/tree/master/packages/scss-config-webpack-plugin/config
    new ScssConfigWebpackPlugin(),
  ],
};


⚠️  Skipping NPM installation
💡  To install the dependencies manually run:

npm install --save-dev  @babel/core @babel/preset-env @babel/preset-react asset-config-webpack-plugin clean-webpack-plugin html-webpack-plugin js-config-webpack-plugin scss-config-webpack-plugin webpack webpack-cli webpack-dev-server

"
`;

exports[`cli it output the correct webpack.config.js content for {"options":["useClean","useCli","useCss","useDevServer","useFonts","useHtml","useImages","useScss","useTs"]} 1`] = `
"
⚠️ Skipping writing webpack.config.js
💡  Add the following code to your webpack.config.js:

const AssetConfigWebpackPlugin = require('asset-config-webpack-plugin');
const CleanWebpackPlugin = require('clean-webpack-plugin');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const ScssConfigWebpackPlugin = require('scss-config-webpack-plugin');
const TsConfigWebpackPlugin = require('ts-config-webpack-plugin');

module.exports = {
  plugins: [
    // File loader configuration for .woff and .woff2 files
    // File loader configuration for .gif .jpg .jpeg .png and .svg files
    // https://github.com/namics/webpack-config-plugins/tree/master/packages/asset-config-webpack-plugin
    new AssetConfigWebpackPlugin(),
    // Cleans the dist folder before the build starts
    new CleanWebpackPlugin(['dist']),
    // Generate a base html file and injects all generated css and js files
    new HtmlWebpackPlugin(),
    // SCSS Configuration for .css .module.css and .scss .module.scss files
    // see https://github.com/namics/webpack-config-plugins/tree/master/packages/scss-config-webpack-plugin/config
    new ScssConfigWebpackPlugin(),
    // Multi threading typescript loader configuration with caching for .ts and .tsx files
    // see https://github.com/namics/webpack-config-plugins/tree/master/packages/ts-config-webpack-plugin/config
    new TsConfigWebpackPlugin(),
  ],
};


⚠️  Skipping NPM installation
💡  To install the dependencies manually run:

npm install --save-dev  asset-config-webpack-plugin clean-webpack-plugin html-webpack-plugin scss-config-webpack-plugin ts-config-webpack-plugin typescript webpack webpack-cli webpack-dev-server

"
`;

exports[`cli it output the correct webpack.config.js content for {"options":["useCli","useJs"]} 1`] = `
"
⚠️ Skipping writing webpack.config.js
💡  Add the following code to your webpack.config.js:

const JsConfigWebpackPlugin = require('js-config-webpack-plugin');

module.exports = {
  plugins: [
    // Multi threading babel loader configuration with caching for .js and .jsx files
    // see https://github.com/namics/webpack-config-plugins/tree/master/packages/js-config-webpack-plugin/config
    new JsConfigWebpackPlugin(),
  ],
};


⚠️  Skipping NPM installation
💡  To install the dependencies manually run:

npm install --save-dev  @babel/core @babel/preset-env @babel/preset-react js-config-webpack-plugin webpack webpack-cli

"
`;

exports[`cli it output the correct webpack.config.js content for {"options":["useCli","useTs"]} 1`] = `
"
⚠️ Skipping writing webpack.config.js
💡  Add the following code to your webpack.config.js:

const TsConfigWebpackPlugin = require('ts-config-webpack-plugin');

module.exports = {
  plugins: [
    // Multi threading typescript loader configuration with caching for .ts and .tsx files
    // see https://github.com/namics/webpack-config-plugins/tree/master/packages/ts-config-webpack-plugin/config
    new TsConfigWebpackPlugin(),
  ],
};


⚠️  Skipping NPM installation
💡  To install the dependencies manually run:

npm install --save-dev  ts-config-webpack-plugin typescript webpack webpack-cli

"
`;

exports[`cli it output the correct webpack.config.js content for {"options":["useDevServer","useJs","useImages","useScss"]} 1`] = `
"
⚠️ Skipping writing webpack.config.js
💡  Add the following code to your webpack.config.js:

const ImagesConfigWebpackPlugin = require('image-config-webpack-plugin');
const JsConfigWebpackPlugin = require('js-config-webpack-plugin');
const ScssConfigWebpackPlugin = require('scss-config-webpack-plugin');

module.exports = {
  plugins: [
    // File loader configuration for .gif .jpg .jpeg .png and .svg files
    // see https://github.com/namics/webpack-config-plugins/tree/master/packages/image-config-webpack-plugin/config
    new ImagesConfigWebpackPlugin(),
    // Multi threading babel loader configuration with caching for .js and .jsx files
    // see https://github.com/namics/webpack-config-plugins/tree/master/packages/js-config-webpack-plugin/config
    new JsConfigWebpackPlugin(),
    // SCSS Configuration for .css .module.css and .scss .module.scss files
    // see https://github.com/namics/webpack-config-plugins/tree/master/packages/scss-config-webpack-plugin/config
    new ScssConfigWebpackPlugin(),
  ],
};


⚠️  Skipping NPM installation
💡  To install the dependencies manually run:

npm install --save-dev  @babel/core @babel/preset-env @babel/preset-react image-config-webpack-plugin js-config-webpack-plugin scss-config-webpack-plugin webpack webpack-dev-server

"
`;

exports[`cli it output the correct webpack.config.js content for {"options":["useDevServer","useJs"]} 1`] = `
"
⚠️ Skipping writing webpack.config.js
💡  Add the following code to your webpack.config.js:

const JsConfigWebpackPlugin = require('js-config-webpack-plugin');

module.exports = {
  plugins: [
    // Multi threading babel loader configuration with caching for .js and .jsx files
    // see https://github.com/namics/webpack-config-plugins/tree/master/packages/js-config-webpack-plugin/config
    new JsConfigWebpackPlugin(),
  ],
};


⚠️  Skipping NPM installation
💡  To install the dependencies manually run:

npm install --save-dev  @babel/core @babel/preset-env @babel/preset-react js-config-webpack-plugin webpack webpack-dev-server

"
`;

exports[`cli it output the correct webpack.config.js content for {"options":["useDevServer","useTs","useImages","useScss"]} 1`] = `
"
⚠️ Skipping writing webpack.config.js
💡  Add the following code to your webpack.config.js:

const ImagesConfigWebpackPlugin = require('image-config-webpack-plugin');
const ScssConfigWebpackPlugin = require('scss-config-webpack-plugin');
const TsConfigWebpackPlugin = require('ts-config-webpack-plugin');

module.exports = {
  plugins: [
    // File loader configuration for .gif .jpg .jpeg .png and .svg files
    // see https://github.com/namics/webpack-config-plugins/tree/master/packages/image-config-webpack-plugin/config
    new ImagesConfigWebpackPlugin(),
    // SCSS Configuration for .css .module.css and .scss .module.scss files
    // see https://github.com/namics/webpack-config-plugins/tree/master/packages/scss-config-webpack-plugin/config
    new ScssConfigWebpackPlugin(),
    // Multi threading typescript loader configuration with caching for .ts and .tsx files
    // see https://github.com/namics/webpack-config-plugins/tree/master/packages/ts-config-webpack-plugin/config
    new TsConfigWebpackPlugin(),
  ],
};


⚠️  Skipping NPM installation
💡  To install the dependencies manually run:

npm install --save-dev  image-config-webpack-plugin scss-config-webpack-plugin ts-config-webpack-plugin typescript webpack webpack-dev-server

"
`;

exports[`cli it output the correct webpack.config.js content for {"options":["useDevServer","useTs"]} 1`] = `
"
⚠️ Skipping writing webpack.config.js
💡  Add the following code to your webpack.config.js:

const TsConfigWebpackPlugin = require('ts-config-webpack-plugin');

module.exports = {
  plugins: [
    // Multi threading typescript loader configuration with caching for .ts and .tsx files
    // see https://github.com/namics/webpack-config-plugins/tree/master/packages/ts-config-webpack-plugin/config
    new TsConfigWebpackPlugin(),
  ],
};


⚠️  Skipping NPM installation
💡  To install the dependencies manually run:

npm install --save-dev  ts-config-webpack-plugin typescript webpack webpack-dev-server

"
`;
